pulsar-broker/src/main/java/org/apache/pulsar/broker/TransactionMetadataStoreService.java

```diff
@@ -133,54 +133,56 @@ public CompletableFuture<Void> handleTcClientConnect(TransactionCoordinatorID tc
                             return;
                         }
 
-                        openTransactionMetadataStore(tcId).thenAccept((store) -> internalPinnedExecutor.execute(() -> {
+                        openTransactionMetadataStore(tcId).thenAccept((store) -> {
                             stores.put(tcId, store);
                             LOG.info("Added new transaction meta store {}", tcId);
-                            long endTime = System.currentTimeMillis() + HANDLE_PENDING_CONNECT_TIME_OUT;
-                            while (true) {
-                                // prevent thread in a busy loop.
-                                if (System.currentTimeMillis() < endTime) {
-                                    CompletableFuture<Void> future = deque.poll();
-                                    if (future != null) {
-                                        // complete queue request future
-                                        future.complete(null);
+                            internalPinnedExecutor.execute(() -> {
+                                long endTime = System.currentTimeMillis() + HANDLE_PENDING_CONNECT_TIME_OUT;
+                                while (true) {
+                                    // prevent thread in a busy loop.
+                                    if (System.currentTimeMillis() < endTime) {
+                                        CompletableFuture<Void> future = deque.poll();
+                                        if (future != null) {
+                                            // complete queue request future
+                                            future.complete(null);
+                                        } else {
+                                            break;
+                                        }
                                     } else {
+                                        deque.clear();
                                         break;
                                     }
-                                } else {
-                                    deque.clear();
-                                    break;
                                 }
-                            }
 
-                            completableFuture.complete(null);
-                            tcLoadSemaphore.release();
-                        })).exceptionally(e -> {
+                                completableFuture.complete(null);
```

Comment:
@congbobo184 @thetumbled Sorry, I didn't get the key point of the problem. The completableFuture is completed by the same thread of executing `stores.put(tcId, store);` Why will we have race condition here? The client-side should send the end transaction command after completing the TC connect stage.

Reply:
There are two thread pool `org.apache.pulsar.transaction.coordinator.impl.MLTransactionMetadataStore#internalPinnedExecutor` and `org.apache.pulsar.broker.TransactionMetadataStoreService#internalPinnedExecutor`. They are different.

Reply:
Yes, but before we complete the `completableFuture` here, we have added the item to the map. Why the subsequent request is not able to get from the map? The subsequent request only happened after the `completableFuture` was done, right? And the map is a `ConcurrentHashMap`, so what is the race condition here? Could you please provide more details about the race condition? How does it happen?

Reply:
The completableFuture return by `openTransactionMetadataStore(tcId)` is completed by following code in thread `MLTransactionMetadataStore#internalPinnedExecutor`.
```
                        **completableFuture.complete(MLTransactionMetadataStore.this);** 
                        recoverTracker.handleCommittingAndAbortingTransaction();
```
Once the completableFuture is completed, `stores.put` will be executed in another thread `TransactionMetadataStoreService#internalPinnedExecutor`.
```
                        openTransactionMetadataStore(tcId).thenAccept((store) -> internalPinnedExecutor.execute(() -> {
                            stores.put(tcId, store);
```
So, we may execute `recoverTracker.handleCommittingAndAbortingTransaction()` before execute `stores.put(tcId, store)`.

Reply:
Essentially this is a circular dependency problem.

`TransactionMetadataStoreService.handleTcClientConnect` -> `MLTransactionMetadataStore.init` -> `TransactionRecoverTracker.handleCommittingAndAbortingTransaction` -> `TransactionMetadataStoreService.handleTcClientConnect`

It looks like we need to ensure some state is changed in `TransactionMetadataStoreService` during init `MLTransactionMetadataStore`.

IMO, we should refactor this part finally to move the `recoverTracker.handleCommittingAndAbortingTransaction();` to the `TransactionMetadataStoreService` to decouple the mutual state dependence.

@congbobo184 @liangyepianzhou WDYT?

It's hard to understand why the map put operation should be executed out of the internalPinnedExecutor while reading the code. This may present challenges for future maintenance.

Reply:
`openTransactionMetadataStore(tcId`) can return `MutablePair<store, recoverTracker>` or after `openTransactionMetadataStore(tcId)` then init return the `recoverTracker`. I prefer to use the second way, in this way the logical more clear, after store init the tracker need to do something to handle the legacy the committing and abort transactions.

Reply:
the first solution is more concise, and i have implement it.

Reply:
yes, in this way can completely decoupled. @codelipenghui WDYT?

Reply:
Sorry, I haven't got the point of `openTransactionMetadataStore(tcId) then init return the recoverTracker`, @congbobo184 Can you share the link?

Reply:
https://github.com/apache/pulsar/blob/3bab0997ffc71973c46abf229f61b6d24b6354e6/pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/MLTransactionMetadataStore.java#L108
change to `public CompletableFuture<TransactionRecoverTracker> init()`

`openTransactionMetadataStore(tcId)` only return the store, 

then the store can add a interface `public CompletableFuture<TransactionRecoverTracker> init()`,
 
`TransactionMetadataStoreService` can invoke the store.init() return TransactionRecoverTracker

init completely, recoverTracker can do
https://github.com/apache/pulsar/blob/3bab0997ffc71973c46abf229f61b6d24b6354e6/pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/TransactionRecoverTracker.java#L47
https://github.com/apache/pulsar/blob/3bab0997ffc71973c46abf229f61b6d24b6354e6/pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/TransactionRecoverTracker.java#L52
in the `TransactionMetadataStoreService`